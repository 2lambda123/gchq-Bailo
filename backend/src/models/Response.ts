import { Document, model, Schema } from 'mongoose'
import MongooseDelete from 'mongoose-delete'

export const Decision = {
  RequestChanges: 'request_changes',
  Approve: 'approve',
  Undo: 'undo',
} as const
export type DecisionKeys = (typeof Decision)[keyof typeof Decision]

export const ResponseKind = {
  Review: 'review',
  Comment: 'comment',
} as const
export type ResponseKindKeys = (typeof ResponseKind)[keyof typeof ResponseKind]

export interface ResponseInterface {
  user: string
  kind: ResponseKindKeys
  role?: string
  decision?: DecisionKeys
  comment?: string

  createdAt: string
  updatedAt: string
}

// The doc type includes all values in the plain interface, as well as all the
// properties and functions that Mongoose provides.  If a function takes in an
// object from Mongoose it should use this interface
export type ResponseDoc = ResponseInterface & Document<any, any, ResponseInterface>

const ResponseSchema = new Schema<ResponseInterface>(
  {
    user: { type: String, required: true },
    kind: { type: String, enum: Object.values(ResponseKind) },
    role: { type: String },
    decision: { type: String, enum: Object.values(Decision) },
    comment: { type: String, required: true },
  },
  {
    timestamps: true,
    collection: 'v2_responses',
  },
)

ResponseSchema.plugin(MongooseDelete, {
  overrideMethods: 'all',
  deletedBy: true,
  deletedByType: String,
})

const ResponseModel = model<ResponseInterface>('v2_response', ResponseSchema)

export default ResponseModel
