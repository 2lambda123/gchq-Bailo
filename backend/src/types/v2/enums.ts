export const SchemaKind = {
  Model: 'model',
  Deployment: 'deployment',
} as const
export type SchemaKindKeys = (typeof SchemaKind)[keyof typeof SchemaKind]

export const GetModelFilters = {
  Mine: 'mine',
} as const

export type GetModelFiltersKeys = (typeof GetModelFilters)[keyof typeof GetModelFilters]

export const ReviewKind = {
  Release: 'release',
  Access: 'access',
} as const
export type ReviewKindKeys = (typeof ReviewKind)[keyof typeof ReviewKind]
