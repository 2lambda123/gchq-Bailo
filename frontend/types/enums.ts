export enum Styling {
  NARROW_WIDTH = '750px',
}
