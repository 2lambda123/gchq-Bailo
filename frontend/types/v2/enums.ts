export enum ApprovalStates {
  Accepted = 'approve',
  RequestChanges = 'request_changes',
}
