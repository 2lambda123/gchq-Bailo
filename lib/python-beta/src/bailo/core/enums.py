from __future__ import annotations

from enum import Enum


class ModelVisibility(str, Enum):
    Private = 'private'
    Public = 'public'

    def __str__(self):
        return str(self.value)

class SchemaKind(str, Enum):
    Model = 'model'
    AccessRequest = 'accessRequest'
