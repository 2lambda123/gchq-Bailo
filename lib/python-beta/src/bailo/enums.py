from enum import Enum

class ModelVisibility(str, Enum):
    Private = 'private'
    Public = 'public'
