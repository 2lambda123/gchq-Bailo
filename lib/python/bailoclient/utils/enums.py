from enum import Enum


class ModelFlavours(Enum):
    H2O = "h2o"
    KERAS = "keras"
    MLEAP = "mleap"
    PYTORCH = "pytorch"
    SKLEARN = "sklearn"
    SPARK = "spark"
    TENSORFLOW = "tensorflow"
    ONNX = "onnx"
    GLUON = "gluon"
    XGBOOST = "xgboost"
    LIGHTGBM = "lightgbm"
    CATBOOST = "catboost"
    SPACY = "spacy"
    FASTAI = "fastai"
    STATSMODELS = "statsmodels"
    PROPHET = "prophet"
