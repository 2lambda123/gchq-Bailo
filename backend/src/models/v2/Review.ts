import { Document, model, Schema } from 'mongoose'
import MongooseDelete from 'mongoose-delete'

import { ReviewKind, ReviewKindKeys } from '../../types/v2/enums.js'

export const Decision = {
  RequestChanges: 'request_changes',
  Approve: 'approve',
} as const
export type DecisionKeys = (typeof Decision)[keyof typeof Decision]

interface ReviewResponse {
  user: string
  decision: DecisionKeys
  comment: string
}

// This interface stores information about the properties on the base object.
// It should be used for plain object representations, e.g. for sending to the
// client.
export interface ReviewInterface {
  semver?: string
  modelId?: string
  kind: ReviewKindKeys

  role: string

  responses: Array<ReviewResponse>

  createdAt: Date
  updatedAt: Date
}

// The doc type includes all values in the plain interface, as well as all the
// properties and functions that Mongoose provides.  If a function takes in an
// object from Mongoose it should use this interface
export type ReviewDoc = ReviewInterface & Document<any, any, ReviewInterface>

const ReviewSchema = new Schema<ReviewInterface>(
  {
    semver: {
      type: String,
      required: function (this: ReviewInterface): boolean {
        return this.kind === ReviewKind.Release
      },
    },
    modelId: { type: String, required: true },
    kind: { type: String, enum: Object.values(ReviewKind), required: true },

    role: { type: String, required: true },

    responses: [
      {
        user: { type: String, required: true },
        decision: { type: String, enum: Object.values(Decision), required: true },
        comment: { type: String, required: false },
      },
    ],
  },
  {
    timestamps: true,
    collection: 'v2_reviews',
  }
)

ReviewSchema.plugin(MongooseDelete, {
  overrideMethods: 'all',
  deletedBy: true,
  deletedByType: String,
})

const ReviewModel = model<ReviewInterface>('v2_Review', ReviewSchema)

export default ReviewModel
