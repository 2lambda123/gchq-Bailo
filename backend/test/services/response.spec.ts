import { describe, expect, test, vi } from 'vitest'

import { Decision } from '../../src/models/Response.js'
import { ReviewDoc } from '../../src/models/Review.js'
import {
  checkAccessRequestsApproved,
  respondToReview,
  sendReviewResponseNotification,
} from '../../src/services/response.js'
import { ReviewKind } from '../../src/types/enums.js'
import { testReleaseReview, testReviewResponse } from '../testUtils/testModels.js'

vi.mock('../../src/connectors/authorisation/index.js')
vi.mock('../../src/connectors/authentication/index.js', async () => ({
  default: { getEntities: vi.fn(() => ['user:test']) },
}))

const responseModelMock = vi.hoisted(() => {
  const obj: any = {}

  obj.aggregate = vi.fn(() => obj)
  obj.match = vi.fn(() => obj)
  obj.sort = vi.fn(() => obj)
  obj.lookup = vi.fn(() => obj)
  obj.append = vi.fn(() => obj)
  obj.find = vi.fn(() => obj)
  obj.findOne = vi.fn(() => obj)
  obj.findOneAndUpdate = vi.fn(() => obj)
  obj.findByIdAndUpdate = vi.fn(() => obj)
  obj.updateOne = vi.fn(() => obj)
  obj.save = vi.fn(() => obj)
  obj.limit = vi.fn(() => obj)
  obj.unwind = vi.fn(() => obj)
  obj.at = vi.fn(() => obj)
  obj.map = vi.fn(() => [])
  obj.filter = vi.fn(() => [])

  const model: any = vi.fn(() => obj)
  Object.assign(model, obj)

  return model
})

vi.mock('../../src/models/Response.js', async () => ({
  ...((await vi.importActual('../../src/models/Response.js')) as object),
  default: responseModelMock,
}))

const smtpMock = vi.hoisted(() => ({
  notifyReviewResponseForAccess: vi.fn(() => Promise.resolve()),
  notifyReviewResponseForRelease: vi.fn(() => Promise.resolve()),
  requestReviewForRelease: vi.fn(() => Promise.resolve()),
  requestReviewForAccessRequest: vi.fn(() => Promise.resolve()),
}))
vi.mock('../../src/services/smtp/smtp.js', async () => smtpMock)

const reviewMock = vi.hoisted(() => ({
  findReviewsForAccessRequests: vi.fn(() => [testReleaseReview]),
  findReviewForResponse: vi.fn(() => testReviewResponse),
}))
vi.mock('../../src/services/review.js', async () => reviewMock)

const modelMock = vi.hoisted(() => ({
  getModelById: vi.fn(),
}))
vi.mock('../../src/services/model.js', async () => modelMock)

const accessRequestServiceMock = vi.hoisted(() => ({
  getAccessRequestById: vi.fn(),
}))
vi.mock('../../src/services/accessRequest.js', async () => accessRequestServiceMock)

const releaseRequestServiceMock = vi.hoisted(() => ({
  getReleaseBySemver: vi.fn(),
}))
vi.mock('../../src/services/release.js', async () => releaseRequestServiceMock)

const logMock = vi.hoisted(() => ({
  info: vi.fn(),
  warn: vi.fn(),
  error: vi.fn(),
}))
vi.mock('../../src/services/log.js', async () => ({
  default: logMock,
}))
const arrayUtilMock = vi.hoisted(() => ({
  asyncFilter: vi.fn(),
}))
vi.mock('../../src/utils/array.js', async () => arrayUtilMock)

const mockWebhookService = vi.hoisted(() => {
  return {
    sendWebhooks: vi.fn(),
  }
})
vi.mock('../../src/services/webhook.js', () => mockWebhookService)

describe('services > review', () => {
  const user: any = { dn: 'test' }

  test('respondToReview > release successful', async () => {
    await respondToReview(
      user,
      'modelId',
      'msro',
      {
        decision: Decision.RequestChanges,
        comment: 'Do better!',
      },
      ReviewKind.Release,
      'semver',
    )

    expect(responseModelMock.match.mock.calls.at(0)).toMatchSnapshot()
    expect(responseModelMock.match.mock.calls.at(1)).toMatchSnapshot()
    expect(mockWebhookService.sendWebhooks).toBeCalled()
  })

  test('respondToReview > access request successful', async () => {
    await respondToReview(
      user,
      'modelId',
      'msro',
      {
        decision: Decision.RequestChanges,
        comment: 'Do better!',
      },
      ReviewKind.Access,
      'accessRequestId',
    )

    expect(responseModelMock.match.mock.calls.at(0)).toMatchSnapshot()
    expect(responseModelMock.match.mock.calls.at(1)).toMatchSnapshot()
  })

  test('respondToReview > access request review  response notification successful', async () => {
    accessRequestServiceMock.getAccessRequestById.mockReturnValueOnce({ createdBy: 'Yellow' })
    await sendReviewResponseNotification(
      { kind: 'access', accessRequestId: 'Hello' } as ReviewDoc,
      testReviewResponse as any,
      user,
    )

    expect(accessRequestServiceMock.getAccessRequestById).toBeCalled()
    expect(smtpMock.notifyReviewResponseForAccess).toBeCalled()
  })

  test('respondToReview > missing access request ID', async () => {
    releaseRequestServiceMock.getReleaseBySemver.mockReturnValueOnce({ createdBy: 'Yellow' })
    const err = Error('test error')
    smtpMock.notifyReviewResponseForAccess.mockImplementationOnce(() => {
      throw err
    })
    await sendReviewResponseNotification({ kind: 'access' } as ReviewDoc, testReviewResponse as any, user)

    expect(logMock.error).toHaveBeenCalledWith(
      { review: { kind: 'access' } },
      'Unable to send notification for review response. Cannot find access request ID.',
    )
  })

  test('respondToReview > release review response notification successful', async () => {
    releaseRequestServiceMock.getReleaseBySemver.mockReturnValueOnce({ createdBy: 'Yellow' })
    await sendReviewResponseNotification(
      { kind: 'release', semver: 'Hello' } as ReviewDoc,
      testReviewResponse as any,
      user,
    )

    expect(releaseRequestServiceMock.getReleaseBySemver).toBeCalled()
    expect(smtpMock.notifyReviewResponseForRelease).toBeCalled()
  })

  test('respondToReview > missing semver', async () => {
    releaseRequestServiceMock.getReleaseBySemver.mockReturnValueOnce({ createdBy: 'Yellow' })
    await sendReviewResponseNotification({ kind: 'release' } as ReviewDoc, testReviewResponse as any, user)

    expect(logMock.error).toHaveBeenCalledWith(
      {
        review: { kind: 'release' },
      },
      'Unable to send notification for review response. Cannot find semver.',
    )
  })

  test('respondToReview > no reviews found', async () => {
    reviewMock.findReviewForResponse.mockRejectedValueOnce('Unable to find Review to respond to')

    const result = respondToReview(
      user,
      'modelId',
      'msro',
      {
        decision: Decision.RequestChanges,
        comment: 'Do better!',
      },
      ReviewKind.Release,
      'semver',
    )

    expect(result).rejects.toThrowError(`Unable to find Review to respond to`)
    expect(responseModelMock.save).not.toBeCalled()
  })

  test('checkAccessRequestsApproved > approved access request exists', async () => {
    reviewMock.findReviewsForAccessRequests.mockReturnValueOnce([{ role: 'msro' }, { role: 'random' }] as any)
    responseModelMock.find.mockReturnValueOnce(['approved'])

    const result = await checkAccessRequestsApproved(['access-1', 'access-2'])

    expect(result).toBe(true)
    expect(reviewMock.findReviewsForAccessRequests.mock.calls).toMatchSnapshot()
  })

  test('checkAccessRequestsApproved > no approved access requests with a required role', async () => {
    reviewMock.findReviewsForAccessRequests.mockReturnValueOnce([{ role: 'random' }] as any)

    const result = await checkAccessRequestsApproved(['access-1', 'access-2'])

    expect(result).toBe(false)
    expect(reviewMock.findReviewsForAccessRequests.mock.calls).toMatchSnapshot()
  })
})
