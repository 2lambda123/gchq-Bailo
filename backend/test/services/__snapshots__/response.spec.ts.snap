// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`services > response > checkAccessRequestsApproved > approved access request exists 1`] = `
[
  [
    [
      "access-1",
      "access-2",
    ],
  ],
]
`;

exports[`services > response > checkAccessRequestsApproved > no approved access requests with a required role 1`] = `
[
  [
    [
      "access-1",
      "access-2",
    ],
  ],
]
`;
